### ISSUE TITLE
Concurrency issue with getCurrentClassLogger() (ObjectGraphScanner.ScanProperties)

### ISSUE THREAD

**Issue by ErikApption** [NLog/NLog#500]
I use the unity framework to create my classes - when this happens, I believe GetCurrentClassLogger() gets called on many classes at the same time. Shouldn't this method be thread safe?
This did not happen as frequently with nlog 2.1 - this issue happens with nlog 3.2.
And frequently I get

```
  Resolving Apption.HPCManagementService.Commands.CopyFolderExecute,(none)
  Calling constructor Apption.HPCManagementService.Commands.CopyFolderExecute(System.IO.DirectoryInfo share, Apption.Wcf.Client.ManagerReference.IManagementService svc)
 ---> System.InvalidOperationException: Collection was modified after the enumerator was instantiated.
   at System.Collections.Generic.Queue`1.Enumerator.MoveNext()
   at NLog.Internal.ObjectGraphScanner.ScanProperties[T](List`1 result, Object o, Int32 level, Dictionary`2 visitedObjects)
   at NLog.Internal.ObjectGraphScanner.ScanProperties[T](List`1 result, Object o, Int32 level, Dictionary`2 visitedObjects)
   at NLog.Internal.ObjectGraphScanner.ScanProperties[T](List`1 result, Object o, Int32 level, Dictionary`2 visitedObjects)
   at NLog.Internal.ObjectGraphScanner.ScanProperties[T](List`1 result, Object o, Int32 level, Dictionary`2 visitedObjects)
   at NLog.Internal.ObjectGraphScanner.ScanProperties[T](List`1 result, Object o, Int32 level, Dictionary`2 visitedObjects)
   at NLog.Internal.ObjectGraphScanner.ScanProperties[T](List`1 result, Object o, Int32 level, Dictionary`2 visitedObjects)
   at NLog.Internal.ObjectGraphScanner.ScanProperties[T](List`1 result, Object o, Int32 level, Dictionary`2 visitedObjects)
   at NLog.Internal.ObjectGraphScanner.ScanProperties[T](List`1 result, Object o, Int32 level, Dictionary`2 visitedObjects)
   at NLog.Internal.ObjectGraphScanner.ScanProperties[T](List`1 result, Object o, Int32 level, Dictionary`2 visitedObjects)
   at NLog.Internal.ObjectGraphScanner.FindReachableObjects[T](Object[] rootObjects)
   at NLog.Internal.TargetWithFilterChain.PrecalculateStackTraceUsage()
   at NLog.LogFactory.GetTargetsByLevelForLogger(String name, IList`1 rules, TargetWithFilterChain[] targetsByLevel, TargetWithFilterChain[] lastTargetsByLevel)
   at NLog.LogFactory.GetTargetsByLevelForLogger(String name, IList`1 rules, TargetWithFilterChain[] targetsByLevel, TargetWithFilterChain[] lastTargetsByLevel)
   at NLog.LogFactory.GetConfigurationForLogger(String name, LoggingConfiguration configuration)
   at NLog.LogFactory.GetLogger(LoggerCacheKey cacheKey)
   at NLog.LogManager.GetCurrentClassLogger()
```

**Comment by Xharze** [comment]
This method should be thread safe. Do you know how to reproduce this issue?

**Comment by Xharze** [comment]
Also what does this CopyFolderExecute do?

**Comment by 304NotModified** [comment]
We could test this with an unit test?

**Comment by Xharze** [comment]
The reason I ask about the method is that I know there is an issue with
methods modifying the targets, that's the collection was modified
exception.

**Comment by ErikApption** [comment]
CopyFolderExecute just copies some files. It doesn't modify the targets.
In the initialization, I do have one small section that does indeed add targets in memory. I tried to create a global lock on LogManager.Configuration with no luck.

I tried recreating a regression test but no luck so far.

**Comment by Xharze** [comment]
Thanks for the reproduction example, but what is the MemoryQueueTarget?

**Comment by ErikApption** [comment]
Here is the MemoryQueueTarget

```csharp
    [Target("Memory")]
    public sealed class MemoryQueueTarget : TargetWithLayout
    {
        private int maxSize;
        public MemoryQueueTarget(int size)
        {
            this.Logs = new Queue<string>();
            this.maxSize = size;
            //this.Logs = new List<string>();
        }

        public Queue<string> Logs { get; private set; }

        protected override void Write(LogEventInfo logEvent)
        {
            string msg = this.Layout.Render(logEvent);
            if (msg.Length > 100)
                msg = msg.Substring(0, 100) + "...";

            this.Logs.Enqueue(msg);
            while (this.Logs.Count > maxSize)
            {
                Logs.Dequeue();
            }
        }
    }
```

**Comment by 304NotModified** [comment]
Hi @ErikApption 

I'm working on this. Any idea if this can be reproduced without Unity?

**Comment by 304NotModified** [comment]
never-mind, this has been reproduced and will be fixed in 4.0 :) (2 June release)
